services/devicepolicy/java/com/android/server/devicepolicy/DevicePolicyManagerService.java

```diff
@@ -21241,6 +21244,53 @@ private void maybeInstallDevicePolicyManagementRoleHolderInUser(int targetUserId
         }
     }
 
+    /**
+     * GrapheneOS Handler to install sandboxed play into managed user profile
+     * if requested by the user, in order to allow DPC apps that require
+     * play services to work normally
+     */
+    private void maybeInstallPlay(int targetUserId) {
+        // TODO: Add option to toggle whether or not play should be installed
+        // TODO: possibly copy permissions from existing install in managing user?
+        // TODO: handle play services not being installed correctly (shouldn't do anything in that case)
+        Slogf.d(LOG_TAG, "Installing play for user " + targetUserId);
+
+        List<String> playPkgList = Arrays.asList(PackageId.GSF_NAME, PackageId.GMS_CORE_NAME, PackageId.PLAY_STORE_NAME);
+
+        try {
+            for (final String playPkg : playPkgList) {
+                if (mIPackageManager.isPackageAvailable(playPkg, targetUserId)) {
+                    Slogf.d(LOG_TAG, "The play package "
+                            + playPkg + " is already installed in "
+                            + "user " + targetUserId);
+                    return;
+                }
+                Slogf.d(LOG_TAG, "Installing play package "
+                      + playPkg + " in user " + targetUserId);
+                mIPackageManager.installExistingPackageAsUser(
+                    playPkg,
+                    targetUserId,
+                    PackageManager.INSTALL_ALL_WHITELIST_RESTRICTED_PERMISSIONS,
+                    PackageManager.INSTALL_REASON_POLICY,
+                    /* whiteListedPermissions= */ null);
+            }
+
+            Slogf.d(LOG_TAG, "Granting REQUEST_INSTALL_PACKAGES to Play Store");
+
+            // We need to grant Play Store "Allow from source" as this is not possible
+            // later if DPC policy blocks user installing apps
+            // If DPC policy blocks this setting later the value in UI will be "Enabled by admin"
+
+            final int vendingUID = mInjector.getPackageManagerInternal().getPackageUid(
+                    PackageId.PLAY_STORE_NAME, 0, targetUserId);
+            mInjector.getAppOpsManager().setMode(AppOpsManager.OP_REQUEST_INSTALL_PACKAGES, vendingUID,
```

Comment:
this can be improved like in `        for (AppOpsManager.PackageOps pkgOps : allVpnOps) {` using packageops helper class, to not directly call getPackageManagerInternal